stateresolution.go

```diff
@@ -0,0 +1,279 @@
+		r.resolvedMembers[stateKey] = nil
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[stateKey] = nil
+	default:
+		panic(fmt.Errorf("Unexpected auth event with type %q", eventType))
+	}
+}
+
+// resolveAndAddAuthBlocks resolves each block of conflicting auth state events in a list of blocks
+// where all the blocks have the same event type.
+// Once every block has been resolved the resulting events are added to the events used for auth checks.
+// This is called once per auth event type and state key pair.
+func (r *stateResolver) resolveAndAddAuthBlocks(blocks [][]Event) {
+	start := len(r.result)
+	for _, block := range blocks {
+		if event := r.resolveAuthBlock(block); event != nil {
+			r.result = append(r.result, *event)
+		}
+	}
+	// Only add the events to the auth events once all of the events with that type have been resolved.
+	// (This is done to avoid the result of state resolution depending on the iteration order)
+	for i := start; i < len(r.result); i++ {
+		r.addAuthEvent(&r.result[i])
+	}
+}
+
+// resolveAuthBlock resolves a block of auth events with the same state key to a single event.
+func (r *stateResolver) resolveAuthBlock(events []Event) *Event {
+	// Sort the events by depth and sha1 of event ID
+	block := make([]conflictedEvent, len(events))
+	for i := range events {
+		event := &events[i]
+		block[i] = conflictedEvent{
+			depth:       event.Depth(),
+			eventIDSHA1: sha1.Sum([]byte(event.EventID())),
+			event:       event,
+		}
+	}
+	sort.Sort(conflictedEventSorter(block))
+
+	// Pick the "oldest" event, that is the one with the lowest depth, as the first candidate.
+	result := block[0].event
+	// Temporarily add the candidate event to the auth events.
+	r.addAuthEvent(result)
+	for i := 1; i < len(block); i++ {
+		event := block[i].event
+		// Check if the next event passes authentication checks against the current candidate.
+		if Allowed(*event, r) == nil {
+			// If the event passes authentication checks pick it as the current candidate.
+			result = event
+			r.addAuthEvent(result)
+		} else {
+			break
+		}
+	}
+	// Discard the event from the auth events.
+	// We'll add it back later when all events of the same type have been resolved.
+	r.removeAuthEvent(result.Type(), *result.StateKey())
+	return result
+}
+
+// resolveNormalBlock resolves a block of normal state events with the same state key to a single event.
+func (r *stateResolver) resolveNormalBlock(events []Event) *Event {
+	// Sort the events by depth and sha1 of event ID
+	block := make([]conflictedEvent, len(events))
+	for i := range events {
+		event := &events[i]
+		block[i] = conflictedEvent{
+			depth:       event.Depth(),
+			eventIDSHA1: sha1.Sum([]byte(event.EventID())),
+			event:       event,
+		}
+	}
+	sort.Sort(conflictedEventSorter(block))
+
+	// Start at the "newest" event, that is the one with the highest depth, and go
+	// backward through the list until we find one that passes authentication checks.
```

Comment:
Interesting. Why do we do this by the newest event but the other stuff was by the oldest event? I'm guessing that it has something to do with "well it doesn't affect the integrity of the room so just let everyone clobber everything"? It feels like a good place to mention the intent here.